### PR TITLE
feat: generate types from OpenAPI schemas

### LINKED ISSUE #32: Generate types from Swagger/OpenAPI schemas
### Describe the feature

A huge benefit of writing your backend endpoints using nuxt is the type inference in `useFetch` and `$fetch`. Not only do you get types for the path, but also the request and response data. You don't get that benefit when using API Party. Without types, it's not much of a party.

Some services provide a [json file](https://swagger.io/resources/open-api/) describing the API, including paths and component models. Some frameworks even provide the functionality by default.

## Configuration

I suggest implementing a config that can load a spec from a url, file, or object.

Here's an example.
```ts
// nuxt.config.ts
export default defineNuxtConfig({
  modules: ['nuxt-api-party'],
  apiParty: {
    endpoints: {
      httpbin: {
        url: "https://httpbin.org",
        // URL to the API spec. Could also be a file path or object containing the spec.
        spec: "https://httpbin.org/spec.json"
      }
    }
  }
})
```
## Implementation

For implementation, I would suggest using [openapi-typescript](https://github.com/drwpow/openapi-typescript/tree/main/packages/openapi-typescript) as an optional peer dependency to generate the types. Some types can probably be borrowed from the `openapi-fetch` package to create the `useXXX()` types.

Note though that the latest version only supports openapi, and not swagger, which httpbin uses. Developers would have to install v5 instead of v6.

### Additional information

- [X] Would you be willing to help implement this feature?
- [X] Can you think of other implementations of this feature?

### Final checks

- [X] Check existing [issues](https://github.com/johannschopplich/nuxt-api-party/issues).

### PR DESCRIPTION
**Pull request by killjoy1221**
### 🔗 Linked Issue

#32 

### ❓ Type of change

- [x] 📖 Documentation (updates to the documentation, readme or JSDoc annotations)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

Types can now be generated for an API if it defines an OpenAPI schema. By specifying the schema on the endpoint as a url, file, schema object, or function returning the schema object, types for the paths and components will be generated in `.nuxt/types/nuxt-api-party-types.d.ts`.

Types can be imported directly via

```ts
import { components, paths } from '#nuxt-api-party/myApi'
```

Resolves #32 

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by johannschopplich** [comment]
Dang, that was fast. 👏 Will take me some time to review. But thanks in advance!

File: src/runtime/types.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+// General purpose types
+export type RequestBody<T> = T extends { requestBody: { content: infer Body } }
+  ?
+    | (Body extends { 'application/octet-stream': any }
+      ? {
+          body: string | Blob
+          headers: { 'content-type': 'application/octet-stream' }
+        }
+      : {})
```
Comment (johannschopplich) [inline src/runtime/types.ts:21]: Quick question: What do you mean by `{}`? 

We shouldn't use `{}` as a type. `{}` actually means "any non-nullish value", according to ESLint.
- If you want a type meaning "any object", you probably want `object` instead.
- If you want a type meaning "any value", you probably want `unknown` instead.
- If you want a type meaning "empty object", you probably want `Record<string, never>` instead.
...

See the [eslint@typescript-eslint/ban-types](https://typescript-eslint.io/rules/ban-types/).

**Review by johannschopplich — CHANGES_REQUESTED** [review]

**Comment by johannschopplich** [comment]
Hey @killjoy1221 I just reviewed most of your code. Looks quite promising. I have refactored some code, to bring it closer to Nuxt standards. For example, instead of relying on `require.resolve`, we can use a Nuxt kit util:
```ts
const hasOpenAPIPkg = await tryResolveModule('openapi-typescript', [nuxt.options.rootDir])
```

But in the end, minor optimizations. I haven't touched your logic a lot. The OpenAPI stuff is complicated and I might get 75 % of it. Looking good tho. I have one question left – see review tab. Then, I will have to take a look at why the tests are failing. 
Looks like we can release soon. Thanks a lot for your efforts!

**Comment by killjoy1221** [comment]
No problem. I always expect a bit of nitpicking when it comes to coding standards.

I agree that the OpenAPI types are complex. I tried to support path parameters via template strings ``` `pet/${petId}` ```, but I could never get the type to resolve properly. Maybe I'll get it later and create a new PR.

**Comment by johannschopplich** [comment]
Yeah, maybe TS can't infer the type from a concatenated string. There is the [`Split` utility from `type-fest`](https://github.com/sindresorhus/type-fest#string), which might help?

Anyway, thanks for your contribution. It's a hell of a lot of work! I will release a new package version later today. Glad you helped. :)

**Comment by johannschopplich** [comment]
@killjoy1221 Nuxt API Party v0.14 just released! 🏁

**Pull request merged**